File: tests/scalers/mssql.test.ts
import test from 'ava'
import * as fs from 'fs'
import * as sh from 'shelljs'
import * as tmp from 'tmp'

const mssqlns = "mssql"
const testns = "mssql-app"
const mssqlName = "mssqlinst"
const password = "Pass@word1"
const hostname = `${mssqlName}.${mssqlns}.svc.cluster.local`
const database = "TestDB"
const sqlConnectionString = `Server=${hostname};Database=${database};User ID=sa;Password=${password};`
const appName = "consumer-app"

const getReplicaCountCommand = `kubectl get deployment.apps/${appName} -n ${testns} -o jsonpath="{.spec.replicas}"`

test.before(t => {
    sh.config.silent = true

    // deploy the mssql container
    sh.exec(`kubectl create namespace ${mssqlns}`)
    const mssqlDeploymentYamlFile = tmp.fileSync()
    fs.writeFileSync(mssqlDeploymentYamlFile.name, mssqlDeploymentYaml)
    t.is(0, sh.exec(`kubectl apply -n ${mssqlns} -f ${mssqlDeploymentYamlFile.name}`).code, 'creating the mssql deployment should work.')

    // wait for the mssql container to be ready
    let readyReplicaCount = '0'
    for (let i = 0; i < 30; i++) {
      readyReplicaCount = sh.exec(`kubectl get deploy/mssql-deployment -n ${mssqlns} -o jsonpath='{.status.readyReplicas}'`).stdout
        if (readyReplicaCount != '1') {
            sh.exec('sleep 2s')
        }
    }
    t.is('1', readyReplicaCount, 'mssql-deployment is not in a ready state!')

    // create the mssql database
    const mssqlPod = sh.exec(`kubectl get pods -n ${mssqlns} -o jsonpath='{.items[0].metadata.name}'`).stdout
    t.not(mssqlPod, '')
    sh.exec(`kubectl exec -n ${mssqlns} ${mssqlPod} -- /opt/mssql-tools/bin/sqlcmd -S . -U sa -P "${password}" -Q "CREATE DATABASE [${database}]"`)

    // create the table that KEDA will monitor for scale decisions
    const createTableSQL = "CREATE TABLE tasks ([id] int identity primary key, [status] varchar(10))"
    sh.exec(`kubectl exec -n ${mssqlns} ${mssqlPod} -- /opt/mssql-tools/bin/sqlcmd -S . -U sa -P "${password}" -d "${database}" -Q "${createTableSQL}"`)

    // deploy the test app
    sh.exec(`kubectl create namespace ${testns}`)
    const testAppYamlFile = tmp.fileSync()
    fs.writeFileSync(testAppYamlFile.name, testAppDeployYaml)
    t.is(0, sh.exec(`kubectl apply -n ${testns} -f ${testAppYamlFile.name}`).code, 'creating the test app deployment should work.')
})

test.serial('Deployment should have 0 replicas on start', t => {
    const replicaCount = sh.exec(getReplicaCountCommand).stdout
    t.is(replicaCount, '0', 'replica count should start out as 0')
})

test.serial(`Deployment should scale to 5 (the max) then back to 0`, t => {
    const jobYamlFile = tmp.fileSync()
    fs.writeFileSync(jobYamlFile.name, insertRecordsJobYaml)
    t.is(0, sh.exec(`kubectl apply -f ${jobYamlFile.name} -n ${testns}`).code, 'creating job should work.')

    const maxReplicaCount = '5'

    let replicaCount = '0'
    for (let i = 0; i < 30 && replicaCount !== maxReplicaCount; i++) {
        replicaCount = sh.exec(getReplicaCountCommand).stdout
        if (replicaCount !== maxReplicaCount) {
            sh.exec('sleep 2s')
        }
    }

    t.is(maxReplicaCount, replicaCount, `Replica count should be ${maxReplicaCount} after 60 seconds`)

    for (let i = 0; i < 36 && replicaCount !== '0'; i++) {
        replicaCount = sh.exec(getReplicaCountCommand).stdout
        if (replicaCount !== '0') {
            sh.exec('sleep 5s')
        }
    }

    t.is('0', replicaCount, 'Replica count should be 0 after 3 minutes')
})

test.after.always.cb('clean up deployment artifacts', t => {
    // delete all the app and job resources
    const resources = [
        'scaledobject.keda.sh/mssql-scaledobject',
        'triggerauthentication.keda.sh/keda-trigger-auth-mssql-secret',
        `deployment.apps/${appName}`,
        'secret/mssql-secrets',
        'job/mssql-producer-job',
    ]

    for (const resource of resources) {
        sh.exec(`kubectl delete ${resource} -n ${testns}`)
    }

    sh.exec(`kubectl delete namespace ${testns}`)

    // uninstall mssql
    sh.exec(`kubectl delete -n ${mssqlns} deploy/mssql`)
    sh.exec(`kubectl delete namespace ${mssqlns}`)

    t.end()
})

const testAppDeployYaml = `apiVersion: v1
kind: Secret
metadata:
  name: mssql-secrets
type: Opaque
stringData:
  mssql-sa-password: ${password}
  mssql-connection-string: ${sqlConnectionString}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: mssql-consumer-worker
  name: ${appName}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: mssql-consumer-worker
  template:
    metadata:
      labels:
        app: mssql-consumer-worker
    spec:
      containers:
      - image: docker.io/cgillum/mssqlscalertest:latest
        imagePullPolicy: Always
        name: mssql-consumer-worker
        args: [consumer]
        env:
          - name: SQL_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: mssql-secrets
                key: mssql-connection-string
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-mssql-secret
spec:
  secretTargetRef:
  - parameter: password
    name: mssql-secrets
    key: mssql-sa-password
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: mssql-scaledobject
spec:
  scaleTargetRef:
    name: ${appName}
  pollingInterval: 5
  cooldownPeriod:  10
  minReplicaCount: 0
  maxReplicaCount: 5
  triggers:
  - type: mssql
    metadata:
      host: "${hostname}"
      port: "1433"
      database: "${database}"
      username: sa
      query: "SELECT COUNT(*) FROM tasks WHERE [status]='running' OR [status]='queued'"
      targetValue: "1" # one replica per row
    authenticationRef:
      name: keda-trigger-auth-mssql-secret`

const insertRecordsJobYaml = `apiVersion: batch/v1
kind: Job
metadata:
  labels:
    app: mssql-producer-job
  name: mssql-producer-job
spec:
  template:
    metadata:
      labels:
        app: mssql-producer-job
    spec:
      containers:
      - image: docker.io/cgillum/mssqlscalertest:latest
        imagePullPolicy: Always
        name: mssql-test-producer
        args: ["producer"]
        env:
          - name: SQL_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: mssql-secrets
                key: mssql-connection-string
      restartPolicy: Never
  backoffLimit: 4`

const mssqlDeploymentYaml = `apiVersion: apps/v1
kind: Deployment
metadata:
  name: mssql-deployment
  labels:
    app: mssql
spec:
  replicas: 1
  selector:
     matchLabels:
       app: mssql
  template:
    metadata:
      labels:
        app: mssql
    spec:
      terminationGracePeriodSeconds: 30
      containers:
      - name: mssql
        image: mcr.microsoft.com/mssql/server:2019-latest
        ports:
        - containerPort: 1433
        env:
        - name: MSSQL_PID
          value: "Developer"
        - name: ACCEPT_EULA
          value: "Y"
        - name: SA_PASSWORD
          value: "${password}"
        readinessProbe:
          exec:
            command:
            - /bin/sh
            - -c
            - "/opt/mssql-tools/bin/sqlcmd -S . -U sa -P '${password}' -Q 'SELECT @@Version'"
---
apiVersion: v1
kind: Service
metadata:
  name: ${mssqlName}
spec:
  selector:
    app: mssql
  ports:
    - protocol: TCP
      port: 1433
      targetPort: 1433
  type: ClusterIP`

File: tests/scalers/mssql/mssql_test.go
//go:build e2e
// +build e2e

package mssql_test

import (
	"fmt"
	"testing"

	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../.env")

var mssqlDrivers = []string{"sqlserver", "azuresql"}

type templateData struct {
	TestNamespace             string
	DeploymentName            string
	ScaledObjectName          string
	TriggerAuthenticationName string
	SecretName                string
	MssqlServerName           string
	MssqlHostname             string
	MssqlPassword             string
	MssqlDatabase             string
	MssqlConnectionString     string
	DriverName                string
	MinReplicaCount           int
	MaxReplicaCount           int
}

const (
	deploymentTemplate = `apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: mssql-consumer-worker
  name: {{.DeploymentName}}
  namespace: {{.TestNamespace}}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: mssql-consumer-worker
  template:
    metadata:
      labels:
        app: mssql-consumer-worker
    spec:
      containers:
      - image: ghcr.io/kedacore/tests-mssql:latest
        imagePullPolicy: Always
        name: mssql-consumer-worker
        command: ["/app"]
        args: ["-mode", "consumer"]
        env:
          - name: SQL_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: {{.SecretName}}
                key: mssql-connection-string
`

	secretTemplate = `apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
type: Opaque
stringData:
  mssql-sa-password: {{.MssqlPassword}}
  mssql-connection-string: {{.MssqlConnectionString}}
`

	triggerAuthenticationTemplate = `apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: {{.TriggerAuthenticationName}}
  namespace: {{.TestNamespace}}
spec:
    secretTargetRef:
    - parameter: password
      name: {{.SecretName}}
      key: mssql-sa-password
`

	scaledObjectTemplate = `apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: {{.ScaledObjectName}}
  namespace: {{.TestNamespace}}
spec:
  scaleTargetRef:
    name: {{.DeploymentName}}
  pollingInterval: 5
  cooldownPeriod:  10
  minReplicaCount: {{.MinReplicaCount}}
  maxReplicaCount: {{.MaxReplicaCount}}
  triggers:
  - type: mssql
    metadata:
      host: {{.MssqlHostname}}
      port: "1433"
      database: {{.MssqlDatabase}}
      username: sa
      driverName: {{.DriverName}}
      query: "SELECT COUNT(*) FROM tasks WHERE [status]='running' OR [status]='queued'"
      targetValue: "1" # one replica per row
      activationTargetValue: "15"
    authenticationRef:
      name: {{.TriggerAuthenticationName}}
`

	mssqlStatefulSetTemplate = `apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: {{.MssqlServerName}}
  namespace: {{.TestNamespace}}
  labels:
    app: mssql
spec:
  replicas: 1
  serviceName: {{.MssqlServerName}}
  selector:
     matchLabels:
       app: mssql
  template:
    metadata:
      labels:
        app: mssql
    spec:
      terminationGracePeriodSeconds: 30
      containers:
      - name: mssql
        image: mcr.microsoft.com/mssql/server:2019-latest
        ports:
        - containerPort: 1433
        env:
        - name: MSSQL_PID
          value: "Developer"
        - name: ACCEPT_EULA
          value: "Y"
        - name: SA_PASSWORD
          value: {{.MssqlPassword}}
        readinessProbe:
          exec:
            command:
            - /bin/sh
            - -c
            - "/opt/mssql-tools18/bin/sqlcmd -S . -C -U sa -P '{{.MssqlPassword}}' -Q 'SELECT @@Version'"
`

	mssqlServiceTemplate = `apiVersion: v1
kind: Service
metadata:
  name: {{.MssqlServerName}}
  namespace: {{.TestNamespace}}
spec:
  selector:
    app: mssql
  ports:
    - protocol: TCP
      port: 1433
      targetPort: 1433
  type: ClusterIP
`

	// inserts 10 records in the table
	insertRecordsJobTemplate1 = `apiVersion: batch/v1
kind: Job
metadata:
  labels:
    app: mssql-producer-job
  name: mssql-producer-job1
  namespace: {{.TestNamespace}}
spec:
  template:
    metadata:
      labels:
        app: mssql-producer-job
    spec:
      containers:
      - image: ghcr.io/kedacore/tests-mssql:latest
        imagePullPolicy: Always
        name: mssql-test-producer
        command: ["/app"]
        args: ["-mode", "producer"]
        env:
          - name: SQL_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: {{.SecretName}}
                key: mssql-connection-string
      restartPolicy: Never
  backoffLimit: 4
`

	// inserts 10 records in the table
	insertRecordsJobTemplate2 = `apiVersion: batch/v1
kind: Job
metadata:
  labels:
    app: mssql-producer-job
  name: mssql-producer-job2
  namespace: {{.TestNamespace}}
spec:
  template:
    metadata:
      labels:
        app: mssql-producer-job
    spec:
      containers:
      - image: ghcr.io/kedacore/tests-mssql:latest
        imagePullPolicy: Always
        name: mssql-test-producer
        command: ["/app"]
        args: ["-mode", "producer"]
        env:
          - name: SQL_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: {{.SecretName}}
                key: mssql-connection-string
      restartPolicy: Never
  backoffLimit: 4
  `
)

func TestMssqlScaler(t *testing.T) {
	for _, driver := range mssqlDrivers {
		t.Run(driver, func(t *testing.T) {
			testMssqlScalerDriver(t, driver)
		})
	}
}

func testMssqlScalerDriver(t *testing.T, driverName string) {
	data := newTemplateData(driverName)

	// Create kubernetes resources for MS SQL server
	kc := GetKubernetesClient(t)
	_, mssqlTemplates := getMssqlTemplateData(data)
	_, templates := getTemplateData(data)

	t.Cleanup(func() {
		DeleteKubernetesResources(t, data.TestNamespace, data, mssqlTemplates)
		DeleteKubernetesResources(t, data.TestNamespace, data, templates)
	})

	CreateKubernetesResources(t, kc, data.TestNamespace, data, mssqlTemplates)

	require.True(t, WaitForStatefulsetReplicaReadyCount(t, kc, data.MssqlServerName, data.TestNamespace, 1, 60, 3),
		"replica count should be %d after 3 minutes", 1)

	createDatabaseCommand := fmt.Sprintf("/opt/mssql-tools18/bin/sqlcmd -S . -C -U sa -P \"%s\" -Q \"CREATE DATABASE [%s]\"", data.MssqlPassword, data.MssqlDatabase)

	mssqlServerPodName := fmt.Sprintf("%s-0", data.MssqlServerName)

	ok, out, errOut, err := WaitForSuccessfulExecCommandOnSpecificPod(t, mssqlServerPodName, data.TestNamespace, createDatabaseCommand, 60, 3)
	require.True(t, ok, "executing a command on MS SQL Pod should work; Output: %s, ErrorOutput: %s, Error: %s", out, errOut, err)

	createTableCommand := fmt.Sprintf("/opt/mssql-tools18/bin/sqlcmd -S . -C -U sa -P \"%s\" -d %s -Q \"CREATE TABLE tasks ([id] int identity primary key, [status] varchar(10))\"",
		data.MssqlPassword, data.MssqlDatabase)
	ok, out, errOut, err = WaitForSuccessfulExecCommandOnSpecificPod(t, mssqlServerPodName, data.TestNamespace, createTableCommand, 60, 3)
	require.True(t, ok, "executing a command on MS SQL Pod should work; Output: %s, ErrorOutput: %s, Error: %s", out, errOut, err)

	// Create kubernetes resources for testing
	KubectlApplyMultipleWithTemplate(t, data, templates)
	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, data.DeploymentName, data.TestNamespace, data.MinReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", data.MinReplicaCount)

	testActivation(t, kc, data)
	testScaleOut(t, kc, data)
	testScaleIn(t, kc, data)
}

// insert 10 records in the table -> activation should not happen (activationTargetValue = 15)
func testActivation(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing activation ---")
	KubectlReplaceWithTemplate(t, data, "insertRecordsJobTemplate1", insertRecordsJobTemplate1)

	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, data.DeploymentName, data.TestNamespace, data.MinReplicaCount, 60)
}

// insert another 10 records in the table, which in total is 20 -> should be scaled up
func testScaleOut(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale out ---")
	KubectlReplaceWithTemplate(t, data, "insertRecordsJobTemplate2", insertRecordsJobTemplate2)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, data.DeploymentName, data.TestNamespace, data.MaxReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", data.MaxReplicaCount)
}

func testScaleIn(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale in ---")

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, data.DeploymentName, data.TestNamespace, data.MinReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", data.MinReplicaCount)
}

func newTemplateData(driverName string) templateData {
	testName := fmt.Sprintf("mssql-%s-test", driverName)

	testNamespace := fmt.Sprintf("%s-ns", testName)
	deploymentName := fmt.Sprintf("%s-deployment", testName)
	scaledObjectName := fmt.Sprintf("%s-so", testName)
	triggerAuthenticationName := fmt.Sprintf("%s-ta", testName)
	secretName := fmt.Sprintf("%s-secret", testName)
	mssqlServerName := fmt.Sprintf("%s-server", testName)
	mssqlPassword := "Pass@word1"
	mssqlDatabase := "TestDB"
	mssqlHostname := fmt.Sprintf("%s.%s.svc.cluster.local", mssqlServerName, testNamespace)
	mssqlConnectionString := fmt.Sprintf(
		"Server=%s;Database=%s;User ID=sa;Password=%s;",
		mssqlHostname, mssqlDatabase, mssqlPassword,
	)

	return templateData{
		TestNamespace:             testNamespace,
		DeploymentName:            deploymentName,
		ScaledObjectName:          scaledObjectName,
		TriggerAuthenticationName: triggerAuthenticationName,
		SecretName:                secretName,
		MssqlServerName:           mssqlServerName,
		MssqlHostname:             mssqlHostname,
		MssqlPassword:             mssqlPassword,
		MssqlDatabase:             mssqlDatabase,
		MssqlConnectionString:     mssqlConnectionString,
		DriverName:                driverName,
		MinReplicaCount:           0,
		MaxReplicaCount:           5,
	}
}

func getMssqlTemplateData(data templateData) (templateData, []Template) {
	return data, []Template{
		{Name: "mssqlStatefulSetTemplate", Config: mssqlStatefulSetTemplate},
		{Name: "mssqlServiceTemplate", Config: mssqlServiceTemplate},
	}
}

func getTemplateData(data templateData) (templateData, []Template) {
	return data, []Template{
		{Name: "secretTemplate", Config: secretTemplate},
		{Name: "deploymentTemplate", Config: deploymentTemplate},
		{Name: "triggerAuthenticationTemplate", Config: triggerAuthenticationTemplate},
		{Name: "scaledObjectTemplate", Config: scaledObjectTemplate},
	}
}
